Assembly references added as dependencies

hmmueller opened the pull request:
Hallo Thomas - siehe Kommentar --> ich werde das irgendwann nächste Woche in IXOS einmal ausprobieren.

AAAAABER: 

(a) Ich hab noch immer und schon wieder keine Ahnung, wie bei mir die Version hochgezählt werden könnte (angezeigt bei Aufruf mit /? wird eine Version 2.1a - keine Ahnung, woher das kommt).

(b) Ich kann die Unittests nicht laufen lassen. 
Der MS-Test-Apparat crasht immer mit Folgendem (auch wenn man lt. diversen Anleitungen den Trace-Level auf 4 stellt, steht im log nicht mehr drin:):

Informational: ------ Discover test started ------
V, 8316, 11, 2014/05/03, 13:44:02.700, 217688945578, devenv.exe, Using effective platform:X86 effective framework:Framework45
Error: Exception has been thrown by the target of an invocation.
Informational: ========== Discover test finished: 0 found (0:00:00,0210012) ==========

R# erkennt die Tests korrekt, aber der Testlauf crasht, weil JetBrains...MSTest.RunContext offenbar die Property "SolutionDirectory" nicht implementiert.

Wie lässt Du die Tests laufen???

hmmueller committed:
Assembly references added as dependencies

thoemmi commented:
Hallo Harald, erstmal alles Gute nachträglich zum Geburtstag :smile: 

(a) die Versionsinfo wird aus Informationen in Git erzeugt, siehe [CreateAssemblyInfoFromGit](https://github.com/thoemmi/CreateAssemblyInfoFromGit). Vor einem Release füge ich ein entsprechendes Tag in Git hinzu.
Leider haben wir die Version fehlerhaft angezeigt: nur Major und Minor Version gefolgt von einem hartcodiertem `a`. Ich habe einen Fix 229beaa347d1ecb5838f0381f582da284cf69e23 committed.

(b) bei mir laufen die Tests durch, sowohl mit Visual Studios Bordmitteln als auch mit ReSharpers Test Runner. Leider sind sie zum Teil rot :worried: 

![image](https://cloud.githubusercontent.com/assets/73690/2872800/c78f373c-d377-11e3-813c-d61a13555790.png)
![image](https://cloud.githubusercontent.com/assets/73690/2872801/d3f9d946-d377-11e3-8f9a-72af3eb4715f.png)

hmmueller commented:
Danke!

... rot ist ja ok, weil ich sozuagen "ins Luftleere" implementiert habe ... dann muss ich eben nachgraben, was da bei mir nicht mag :-(

 

Grüße

Harald

 

Gesendet: Sonntag, 04. Mai 2014 um 12:38 UhrVon: "Thomas Freudenberg" notifications@github.comAn: thoemmi/NDepCheck NDepCheck@noreply.github.comCc: "Harald M." harald_m_mueller@gmx.deBetreff: Re: [NDepCheck] Assembly references added as dependencies (#10)

Hallo Harald, erstmal alles gute nachträglich zum Geburtstag 

zu a) die Versionsinfo wird aus Informationen in Git erzeugt, siehe CreateAssemblyInfoFromGit. Leider haben wir sie falsch angezeigt, ich habe einen Fix 229beaa committed.

zu b) bei mir laufen die Tests durch, sowohl mit Visual Studios Bordmitteln als auch mit ReSharpers Test Runner. Leider sind sie zum Teil rot 

—
Reply to this email directly or view it on GitHub.

hmmueller committed:
Assembly references added as dependencies / -a option for ignoring added, and repaired tests

hmmueller commented:
So ... nun alles in Ordnung gebracht (schuld war ich ... mein Laptop: Version von R# und VS2012 war nicht aktuell ... Sorry).

**Zur Änderung**: Sie ist _nicht_ aufwärtskompatibel, weil die neuen Assembly-Dependencies per Default mitgecheckt werden - daher waren auch mehrere Tests rot. Es gibt eine neue Option -a, mit der man die Prüfung abschalten kann; aber die bevorzugte Idee ist,

```
assembly:** ---> assembly:**
```

explizit reinzuschreiben, wenn einen diese Abhängigkeiten nicht interessieren ...
a) Lässt Du das durchgehen, oder muss es aufwärtskompatibel sein = -a muss die Prüfung der Assembly-Abhängigkeiten _einschalten_?
b) Die Abhängigkeiten werden in der Form 

assembly:_Name_/_Version_
assembly:_Name_/_Version_::_Culture_ (wenn Culture vorhanden)

erzeugt. Ist das ok? (mehr oder weniger frei erfunden - eigentlich interessiert i.d.R. wohl nur der Name; das PublicKeyToken doch wohl nie ...).

thoemmi commented:
> **Zur Änderung:** Sie ist nicht aufwärtskompatibel, weil die neuen Assembly-Dependencies per Default mitgecheckt werden - daher waren auch mehrere Tests rot

Hmm, das finde ich etwas suboptimal, aber anders lässt sich das wohl nicht managen: per default ist nichts erlaubt, und man muss alles explizit erlauben. In diesem Fall müssten wir die Major Version auf **3** anheben, weil das ein Breaking Change ist.

> assembly:_Name/Version::Culture_ (wenn Culture vorhanden)

Ist die Culture wirklich notwendig? Ich fände es merkwürdig, wenn es Abhängigkeiten auf spezielle Cultures gäbe. Andererseits kann man damit festlegen, dass nur von der _neutralen_ Culture abhängig sein darf. Da der Part aber optional ist, habe ich da kein Problem mit.

Hast Du die Tests grün bekommen?

hmmueller commented:
a) "Nicht aufwärtskompatibel" - ich meine, dass das Eintragen von

      assembly:*\* ---> assembly:**

leicht genug ist, um diesen Breaking Change zu verantworten ... da kann man aber anderer Meinung sein; und wenn Du das noch einmal anzweifelst, dann dreh ich die Semantik von /a um ==> "prüfe auch assembly-Abhängigkeiten", sodass die Änderung aufwärtskompatibel ist.

 

b) Culture ist überhaupt nicht nötig. Es sind nur 4 "Slots" frei in der Dependency, und da hab ich halt einfach drei davon gefüllt ...

 

c) Tests sind grün - waren zwei fehlende Updates (VS, R#) bei mir - mea culpa!

 

Grüße

Harald

 

Gesendet: Sonntag, 11. Mai 2014 um 22:53 UhrVon: "Thomas Freudenberg" notifications@github.comAn: thoemmi/NDepCheck NDepCheck@noreply.github.comCc: "Harald M." harald_m_mueller@gmx.deBetreff: Re: [NDepCheck] Assembly references added as dependencies (#10)

Zur Änderung: Sie ist nicht aufwärtskompatibel, weil die neuen Assembly-Dependencies per Default mitgecheckt werden - daher waren auch mehrere Tests rot

Hmm, das finde ich etwas suboptimal, aber anders lässt sich das wohl nicht managen: per default ist nichts erlaubt, und man muss alles explizit erlauben. In diesem Fall müssten wir die Major Version auf 3 anheben, weil das ein Breaking Change ist.

assembly:Name/Version::Culture (wenn Culture vorhanden)

Ist die Culture wirklich notwendig? Ich fände es merkwürdig, wenn es Abhängigkeiten auf spezielle Cultures gäbe. Andererseits kann man damit festlegen, dass nur von der neutralen Culture abhängig sein darf. Da der Part aber optional ist, habe ich da kein Problem mit.

Hast Du die Tests grün bekommen?

—
Reply to this email directly or view it on GitHub.

ilkerde commented:
Hallo Harald, Thomas,

Ihr erlaubt mir hoffentlich, dass ich meine Ansicht kurz in die Diskussion einbringen darf. Es geht auch schnell, denn ich habe nur eine Anmerkung zu (a).

Thomas braucht den Bruch der Aufwärtskompatibilität nicht nochmals anzuzweifeln, denn das übernehme ich schon. Im Besonderen finde ich die angedrohte Änderung der Semantik von `/a` sehr charmant. Sie orientiert sich an der Intuition, dass zum Kommando hinzugefügte Schalter den Wirkungsbereich erweitern.

Zugegeben,

```
assembly:** ---> assembly:**
```

ist kein großer Hub - _geschenkt_. Allerdings hebelt es meiner Meinung nach auch ein wenig den Eigencharakter von NDepCheck aus. Darüber hinaus war ich immer der Auffassung, dass gerade die Prüfung auf Nicht-Assembly-Ebene die _Kerneigenschaft_ des Werkzeuges ist. Ein Schalter, der _optional_ die Prüfung auf Assemblies erweitert, träge zu diesem Bild der Kernkompetenzen bei.

Soweit mein unnötiges und überflüssiges Gequatsche. Euch eine gute Zeit und hoffentlich auf Bald!

Besten Gruß,
Ilker

hmmueller commented:
Nix überflüssig. Ich hab eh ein schlechtes Gewissen. Im Bus ist mir aber eingefallen, dass man es auch so machen könnte:

 

Wenn in einem .dep-File eine Regel vorkommt, die "assembly:" enthält, dann wird auch die Assembly-Dependency-Prüfung in Betrieb genommen; sonst nicht.

 

Vorteil: Aufwärtskompatibel; UND keine Änderung an Aufrufen, wenn man doch Assembly-Dependencies prüfen will.

Nachteil: Man kann Verwendungen konstruieren, wo das nicht geht - wenn z.B. eine Assembly-Dep so definiert wird:

 

[a][s][s][e][m][b][l][y]:*\* ---> [a][s][s][e][m][b][l][y]:mscorlib

 

Wer das tut, ist selber schuld. Aber: Man kann dafür doch die /a-Option implementieren, die dann das "intelligente heuristische Verfahren" übersteuert (=Assembly-Dependency-Prüfung einschaltet).

Noch ein Nachteil: Komplexität des Verhaltens steigt. Da hab ich in unserer Welt kein wirklich schlechtes Gewissen mehr, solang sich's in dieser Größenordnung abspielt ...

 

Ich hoff, sonst ist alles ok bei Euch!!!

Harald

 

Gesendet: Montag, 12. Mai 2014 um 20:10 UhrVon: ilkerde notifications@github.comAn: thoemmi/NDepCheck NDepCheck@noreply.github.comCc: "Harald M." harald_m_mueller@gmx.deBetreff: Re: [NDepCheck] Assembly references added as dependencies (#10)

Hallo Harald, Thomas,

Ihr erlaubt mir hoffentlich, dass ich meine Ansicht kurz in die Diskussion einbringen darf. Es geht auch schnell, denn ich habe nur eine Anmerkung zu (a).

Thomas braucht den Bruch der Aufwärtskompatibilität nicht nochmals anzuzweifeln, denn das übernehme ich schon. Im Besonderen finde ich die angedrohte Änderung der Semantik von /a sehr charmant. Sie orientiert sich an der Intuition, dass zum Kommando hinzugefügte Schalter den Wirkungsbereich erweitern.

Zugegeben,

assembly:*\* ---> assembly:**

ist kein großer Hub - geschenkt. Allerdings hebelt es meiner Meinung nach auch ein wenig den Eigencharakter von NDepCheck aus. Darüber hinaus war ich immer der Auffassung, dass gerade die Prüfung auf Nicht-Assembly-Ebene die Kerneigenschaft des Werkzeuges ist. Ein Schalter, der optional die Prüfung auf Assemblies erweitert, träge zu diesem Bild der Kernkompetenzen bei.

Soweit mein unnötiges und überflüssiges Gequatsche. Euch eine gute Zeit und hoffentlich auf Bald!

Besten Gruß,
Ilker

—
Reply to this email directly or view it on GitHub.

thoemmi commented:
Klingt gut, @hmmueller . Kannst Du diesen PR entsprechend ergänzen?

hmmueller commented:
Ja mach ich ... leider erst am Freitag (Chorproben und andere musikalische Termine bremsen mich grad aus).

 

Gesendet: Donnerstag, 15. Mai 2014 um 08:44 UhrVon: "Thomas Freudenberg" notifications@github.comAn: thoemmi/NDepCheck NDepCheck@noreply.github.comCc: "Harald M." harald_m_mueller@gmx.deBetreff: Re: [NDepCheck] Assembly references added as dependencies (#10)

Klingt gut, @hmmueller . Kannst Du diesen PR entsprechend ergänzen?

—
Reply to this email directly or view it on GitHub.

hmmueller committed:
Redo of assembly dependencies
* heuristics: If no rules starting with "assembly:' are present, no assembly rule checking is done; otherwise it is done.
* "flag for speed": -a = CheckOnlyAssemblyDependencies
* 3 unit tests + (more) correct patterns for assembly rules (the version at the end needed special handling)

hmmueller committed:
Version 3 raw commit

hmmueller committed:
Added [NotNull], [CanBeNull] liberally; removed auto check before printing dependcies ist most cases

hmmueller committed:
Correction for & in names; output of include file hierarchy for some error reports; and a little better reduction output for notok example

hmmueller committed:
This is version 3.0, for sure (even though there are a few "??" strings in there ...)

hmmueller committed:
Better errors messages; rules can omit previous left side; removed (currently not working) Nuget dependencies

hmmueller committed:
Handle missing rulefiles better

hmmueller committed:
Added "recursive" rules

hmmueller committed:
Many small improvements and bug fixes

hmmueller committed:
Error message for missing file improved

hmmueller committed:
Memory handling improved; -@ option repaired; macros support itemtypes; a little improvement for "reading file" output

hmmueller committed:
Commit changed handling of @ options; no creation of "one-sided dependencies" (where usedItem is null)

hmmueller committed:
Some attempts at runtime and memory optimizations

hmmueller committed:
And agian better option handling; graphing with explicit types (-y option [chatty is now -w]); etc.

hmmueller committed:
Lifted to .Net 4.5; fixed (but not completely understood) AttributeTest

hmmueller committed:
Reset corrected; Info outputs on -c, -b, -z

hmmueller committed:
Bugfix in DipReader; experiments with drawing

hmmueller committed:
Better/correct handling of empty and ** matching, especially with groups; raw snippets of a future graphic renderer

hmmueller committed:
Merge branch 'master' of https://github.com/hmmueller/NDepCheck.git

hmmueller committed:
Much better definition for projections (was "graphing"); some steps towards a rendering API.

hmmueller committed:
Rendering improved considerably

hmmueller committed:
Renderingtests organized, backgroundcolor override, started with an IXOS application renderer

hmmueller committed:
Refactored Rendering (new -g option for plugins not yet tested)

hmmueller committed:
Rendering via -r option works

hmmueller committed:
Generalized VariableVector to BoundedVector

hmmueller committed:
Added -g option for testing renderers; some more small improvements to rendering, refacotrings/renamings.

hmmueller committed:
IBox has properties for all 8 sector points; names possible for all vectors; Boxes use full diagonal instead of half diagonal; IXOSRenderer has first testdata;

hmmueller committed:
Vertical placement of texts a little better (but still wrong); vectors with more debug output; filename output of gif tests corrected; some more tests with stacked boxes; continued with IXOSRenderer work.

hmmueller committed:
Debug output removed which disturbs caching; text placement implemented (works for Boxes; works ofr non-inclined text on Lines); removed Store class (replaced by Item.DynamicData)

hmmueller committed:
Text placement finally more or less ok

hmmueller committed:
Negative textLocation for absolute positioning of texts on lines; IXOSRenderer creates useful output for test data

hmmueller committed:
Coordinate computation completely rewritten to a simple constraint solver (that needs quite a bit of help to place the things ... that's why it's "simple")

hmmueller committed:
Extremely extended option handling ... must be documented, otherwise I will forget what it does soon ...

hmmueller committed:
NotOkExample handling simplified (is now only a string); -r and -g allow . as assembly, meaning NDepCheck itself; DipWriter is now an IDependencyWriter; previous two items make -c de facto useless

hmmueller committed:
IXOSRenderer moved from test assembly to NDepCheck and renamed to "ModulesAndInterfacesRenderer"; support for extendible options

hmmueller committed:
Added help text for all renderers

hmmueller committed:
Corrected {...} handling in @ files; and added an error message in Projection

hmmueller committed:
Dependencies count ? and ! separately; rendering does dependency checking; DIP format changed: Use => (instead of ->); and output ? and ! cts separately (not yet in MatrixRenderers)

hmmueller committed:
TODOs added in Program; GIF-Renderer outputs number of outgoing, cycling, incoming dependencies

hmmueller committed:
Graphics renderer: Multiline text handled somewhat better; and limited output on no solution.

hmmueller committed:
Clickable edgeInfo in GraphicsRenderer implementiert

hmmueller committed:
MatrixGraphicsRenderer added; Renderer log writing to output file

hmmueller committed:
Usagetype added;

hmmueller committed:
Option handling and internal state handling totally new ... not much else can be said. Reason: Clear separation of engine options nd plugin options everywhere.

hmmueller committed:
A few renamings ("fileName"); removing some dead code/files; starting with a WebServer.

hmmueller committed:
Projection can distinguish left and right; full (but untested) code for WebServer

hmmueller committed:
Render methods without items; DipReader with standard $ lines for ItemTypes; hexChar support in projections for arbitrary chars; -o files support define expansion; option handling for tight options (-xVALUE) corrected; help returns 0 (instead of 1) in Main

hmmueller committed:
Masterfile handling of renderers simplified; first working version of web server; first version of -l (checks only for presence of master file, no teim stamp comparison); corrected some handling of relative paths to absolute paths

hmmueller committed:
Order in Item; AddTopologicalOrder first version;

hmmueller committed:
ViolationsChecker renamed to Check; Projector renamed to Project; AddTopologicalOrder renamed to AddOrder; AddRevere and ResetCouns transformers added.

hmmueller committed:
Added Remove; ResetCounts renamed to Set

hmmueller committed:
Renamed transformers - ...Deps for dependency transformers, ...Items for item transformers; added raw implementation of RemoveItems;

hmmueller committed:
MatrixDictionary - renames to make code understandable; ProjectItems - Option for back projection prepared; RemoveDeps - -s to -i; RemoveItems - correctly implemented; AddReverseDeps - renamed; WebServer & Program: Prepared interactive mode

hmmueller committed:
Dependency - Usage is now a string set; DipReader - removed dependency on DipReaderFactory (necessary for impl of back projecting); GraphicsRenderer - int options have extract method in Options; FindCycleDeps - first version of simple cycle finder; ProjectItems - implemented back projection (UNTESTED); RemoveEdges/Items/FindCycles/SetDeps: -m option now set-based; AddReverseDeps: -m removes usage, -u adds new one

hmmueller committed:
Option system rewritten; and all program options replaced. Still missing: Options of all plugins.
Moreoever, File Watching written anew - but not at all tested.

hmmueller committed:
Changed a number of options to the new option concept (2 letters for abbr; long names supported).

hmmueller committed:
Patterns - inheritance -> delegation; Markers and MarkerPattern introduced; ProjectItems rewritten to new Options.

hmmueller committed:
CheckDeps refactored to use new Options

hmmueller committed:
A few methods are now private in ItemPattern

hmmueller committed:
ItemMatch and ItemPattern - itemType is optional (if null, it is not compared); FindCycleDeps - uses new options

hmmueller committed:
Option handling improved - required and default information added; MarkItems tests;

hmmueller committed:
IsInnerremoved - should be done with specific markers. Most old transformations removed. ItemMatch.Matches() refactored; ignoreCase also sent to renderers

hmmueller committed:
MarkMinimalCut - first implementation

hmmueller committed:
MarkMinimalCut - big fixes and more tests; a few GetHashCode & Equals methods are no [DebuggerHidden]

hmmueller committed:
MarkMinimalCut - option to mark source-side items (necessary if dependency cut is empty)

hmmueller committed:
Interactive handling much improved; help with filter

hmmueller committed:
Help handling improved, some help texts added

hmmueller committed:
Added ICalculator - at the moment, only to compute a hashcode for short filenames; added comment handling and variable expansion to almost all file reads (required passing GlobalCntext around); FindCycleDeps - repaired buggy algorithm; added about 20 new options, like GC and CD and ...

hmmueller committed:
Commented DynamicData (I suspected debugger&performance problems - but they did not vanish; still, replacement with simple Dictionaries reduces complexity); AddAssociativeDeps - first impl; (hopefully) better guessing of ItemTypes

hmmueller committed:
Corrected memory problem in ConstraintSolver (thanks to Robert for nudging me in the right direction ...); size bug fixed in MatrixGraphicsRenderer [but I do not see why some arrows do not get their point ...]

hmmueller committed:
CheckDeps - corrected  filename handling; AbstractTransformer - allow variables to be set to same value multiple times; AbstractLogger - wrong message for questionable dependencies corrected; InputOption - small abstraction; Program - V.3.51

hmmueller committed:
GetHash - no negative hashes (could be confused with options); AbstractDependencyReader - caching not in reader (is created anew with each file option), but in input context; GraphicsRenderer - options for HTML added; ItemAction - Regex corrected; ModifyItems - idiotic copy/paste bug repaired; GlobalContext - inputContexts in Dictionary;

hmmueller committed:
Help for markers added; markers on AddAssociativeEdges are "combined"=concatenated if they have / in them; some snippets of a "drill graph" example included ...

hmmueller committed:
Bugfix for ItemTail; and better output for items not matching their type

hmmueller committed:
Fixed reset bugs

hmmueller committed:
Version = 3.53

hmmueller committed:
ItemPattern - a pattern passed in always wins (and not the hint); GlobalContext - ResetAll (-dr) also resets all variables; ItemType - generic types are compared only by their field count;

hmmueller committed:
Program - -rd  with - behavior; Assembly...Reader resolves to .dll/.exe directory

hmmueller committed:
Merge branch 'master' of https://github.com/hmmueller/NDepCheck.git

hmmueller committed:
AddItemOrder - 4 OrderBy options added; ProjectItems - supports \> for Order property; Program - /? shows plugin help if no other matching help found;

hmmueller committed:
ItemType - next try of working initialization; a few tiny messages changed.

hmmueller committed:
Added small framework for self-optimizing projectors; and implemented a simple self-optimizing projectors selecting a reduced set of projections by looking at the first character of each item value to be matched.

hmmueller committed:
ProjectItems - added SelfOrganizingPrefixTrieProjector, with only a small simple test.

hmmueller committed:
Program - corrected -dc and -db options

hmmueller committed:
Option - corrected option handling for {...} once more ... ProjectItems - improved code for Cost, but not yet finished

hmmueller committed:
ProjectItems - tuning the projectors: Better (or rather correct) computing of efficiency (CostPerProjection); much better inheritance hierarchy; outside control of strategy, at least for testing;

hmmueller committed:
ProjectItems - bugfixes in PrefixTrieProjector; some performance improvements; ItemPattern - Equals/StartsWith/Contains/EndsWith-Matchers now also work with outer parentheses

hmmueller committed:
Test...FirstLetterProjector - test fix

hmmueller committed:
Program - -dc now correctly starts processes with parameters; -dt (-do-time) option added; Option - provide asslowed options if option error

hmmueller committed:
DipReader and DipWriter - a little more output about read and writen dependencies; GlobalContext - modified output of current dependencies a little

hmmueller committed:
Merge branch 'master' of https://github.com/hmmueller/NDepCheck.git

hmmueller committed:
Option - ProcessTextInner pushes inner error messages upwards (makes unitetsting somewhat less awkward); ProjectItems - repaired idiotic bug with regex checking (forgot about |), and added unitest for it!; DipWriter - NDepCheck version is written into file

hmmueller committed:
Merge branch 'master' of https://github.com/hmmueller/NDepCheck

hmmueller committed:
Program - -ls option added; ProjectItems - add a littel output

hmmueller committed:
RuleViolationRenderer, Dependency - logging, source info in bad/questionable messages; Projection, ProjectItems - match count logging; Program - -do-define-default option added, bug with log level setting fixed; GlobalContext - bug with dependency duplication after transform fixed

hmmueller committed:
All transformers - FinishTransform renamed to AfterAllTransforms; Program, GlobalContext and many other places - ExpandDefines is now done on each arg and its options just before execution => standard variable assignment semantics for -dd; Program - -fp option added; GlobalContext - removed unused properties ShowUnused*;

hmmueller committed:
All - GlobalContext parameters uniformly named globalContext; MinimalCutDpes - is now also an ITransformer!; Program - -lr option; ProjectItems - statistics at end; CheckDeps - improved statistics output

hmmueller committed:
AbstractTransformer - forceLoad erroneously removed reinserted; all Transformders with Configure - force reload if confiugratoin is inlined; Program - V.3.59

hmmueller committed:
Program - options help corrected

hmmueller committed:
ITransformer - clarification for a parameter; Program - bug in -dr corrected, q can now also be spelled as -q or -quit or -exit;

hmmueller committed:
Program and Option - corrected option handling problems with -ii, -id, and -?

hmmueller committed:
RuleViolationsRenderer - output improved a little

hmmueller committed:
ValueFrame - names of values are checked (only uppercase A-Z letters ... should be some more language-agnositc pattern ...); ModifyDeps - improved help; GlobalContext - -iw produces more sensible output

hmmueller committed:
Program - V.3.62

hmmueller committed:
AbstractLogger - violations are correctly colored again in ConsoleLogger.

hmmueller committed:
Program - options also expand hex chars (useful for paths with spaces, like "C:\Program Files"); a little more output for -dc;

hmmueller committed:
V.3.63

hmmueller committed:
Beautifications

hmmueller committed:
INode and TestNode abstractions killed - we are dealing solely in Items

hmmueller committed:
IEdge killed - there are only "Dependency" objects (and "Item" objects)

hmmueller committed:
V.3.64

hmmueller committed:
DipReader - ? matching added

hmmueller committed:
AddTransitiveDeps - completely wrong algorithm repaired, e.g. by adding a simple unit test; ItemType - guessing removed; now, for unknown ItemTypes, we always use a GenericType; which is, in addition, more tolerant with field numbers.

hmmueller committed:
ItemWriter - new writer that writes only (matching) items; ItemAction - also "keep" allowed; * - renames of "edge" to "dependency"; ModifyItems - repaired wrong (old tricky) handling of -ml option; Program - V.3.66

hmmueller committed:
ItemType - can now hold a flag for ignorecase and read and write it from a .dip file (+ after name)

hmmueller committed:
DotNetAssemblyReader - various markers on items; DipWriter - markers are persistent; ItemPattern - named fields in pattern; GlobalContext - LogItemCount and LogDependencyCount count as transformations;

hmmueller committed:
V.3.67

hmmueller committed:
Added MarkDeps

hmmueller committed:
TrieProjector - minimal beautification

hmmueller committed:
ProjectItems - a little more beautification

hmmueller committed:
DotNetAssemblyReader - tries to resolve all types; and adds insanely many markers ... too many, probably.

hmmueller committed:
GlobalContext - default for -do-time is 10s (that is about the time after which "waiting" starts); Option - <. and .> can be used instead of { and }, which is helpful on German keyboards.

hmmueller committed:
Test fixes; Program - V.3.68

hmmueller committed:
DotNetAssemblyDependencyReader - removed interning of RawItems - much too slow. DependencyMatch, Item - added help; ItemPattern - small Trim() fix; Program - V.3.69

hmmueller committed:
GlobalContext - LogAboutNItems added; Transformers - using ItemDependencyItemMatch at many places; "Edge" -> "Dependency" at a few places

hmmueller committed:
DependencyMatch, DependencyPattern - renames; Program V.3.70

hmmueller committed:
Transformers - support -dm and -nm wherever possible

hmmueller committed:
...

hmmueller committed:
Additional changes after merge

hmmueller committed:
A few Transformers - MarkerToAddOption renamed to AddMarkerOption, with option changed from -ma to -am; KeepOnlyDeps added; DependencyPattern - input format checking added; Option - IsHelpOption added; GlobalContext - Plugin methods all accept help option; Program - internal transformers can be called without -tf option

hmmueller committed:
Merge mit Main

hmmueller committed:
TestPatternsAndItems - testfix

hmmueller committed:
ItemWriter - added many options

hmmueller committed:
ModulesAndInterfacesRenderer - test data beautified

hmmueller committed:
DependencyMatchOptions - factored out of 5 transformers; AddTransitiveDeps - adds default marker if non provided;

hmmueller committed:
Merge branch 'master' of https://github.com/hmmueller/NDepCheck.git

hmmueller committed:
ItemWriter - suppress items with in/outdegree zero; AddReverse/TransitiveDeps - more output about created/removed dependencies; ObjectWIthMarkers - output about default marker

hmmueller committed:
DipWriter - filtering on output; Program - 3.73

hmmueller committed:
* - RemoveMarkers works now with patterns; DotNetAssemblyDependencyReader - reader markers are prefixed with _;

hmmueller committed:
Item - separate Order field removed, this is now handled by markers

hmmueller committed:
Items & Dependencies - first part of rewrite to immutable Items and Dependencies

hmmueller committed:
"Eat your own dogfood" - the build now contains a call to NDepCheck to check first rules on namespaces and assemblies (this loop is definitely questionable; but a nice example of "higher architecture" in this small project) ; * - moved many things so that static dependencies are nicer, see NDepCheck.nd; Program - V.3.74

hmmueller committed:
Beautifications - dead code removal, moved renderers to two subdirectories; Gibraltar - link to source added

hmmueller committed:
Program - corrected exit code handling; NDepCheck.nd - corrected dependencies for Program after namepsace refactoring; CheckDeps - small message inprovement

hmmueller committed:
Two big changes:
(A) DependencyRule almost fully replaced with Dependency Match - now, also markers can be used in CheckDeps.
(B) DoNetAssemblyReader returns much more information, and - I think - in a somewhat more intuitive way (e.g. some type information together with fields and properties).

The "eat your own dogfood" file (NDepCheck.nd) now contains an example using (A) and (B) for checking the immutability of the ReadOnly* classes.

AbstractMarkerSet - performance hook to subclasses; IMatcher - only one matching method ("Matches"), IsMatch removed; * - estimatedGroupCount -> upperBoundOfGroupCount; Program - V.3.75

hmmueller committed:
RuleViolationWriter, DependencyRuleGroup - small improvements in messages; NDepCheck.nd - added to project; removed 2 old files

hmmueller committed:
Matchers - attempt to save a little time;

hmmueller committed:
GlobalContext - removed commented dead code; FullDotNetAssemblyReader - hardwired resolver path to Silverlight; RuleViolationsWriter - slightly better output format;

hmmueller committed:
Redoing previous two changes (couldn't sync)

hmmueller committed:
AbstractItem - is now fully generic, no dependency on Item any longer; MarkMinimalCut - inner class Edge renamed to EdgeWithFlow

hmmueller committed:
DependencyRuleGroup - fixed iteration bug in checking

hmmueller committed:
PathWriter - first incomplete implementation

hmmueller committed:
Merging different NDpeCheck run results

hmmueller committed:
TestWriters - Removed dead commented code

hmmueller committed:
Program, GlobalContext - fixed Options issue with filenames; DependencyRulGroup - fixed bug with ===> rules

hmmueller committed:
ItemPattern - allow empty unnamed fields in named ItemPatterns.

hmmueller committed:
ItemMatch - ~ prefix to invert match success; MatchResult as return type of matches; PathWriter - first implementation of tree output and list output; DependencyRuleGroup - filter is now full dependency match;

hmmueller committed:
AbstractMarkerSet - default markers get a leading .; MarkerMatch - by default, markers with leading . are not matched (default markers added by transformers); StartsWithMatcher - leading dot remains in place, e.g. for .** pattern; Item - CollectIncoming/OutgoingDependenciesMap returns Dict<Item,Dependency[]>; PathWriter - first ok implementation;

hmmueller committed:
DipWriter - corrected options presentation; GlobalContext - corrected count computation in LogAboutNDependencies

hmmueller committed:
Commit after Merge

hmmueller committed:
ItemPattern - "anywhere" patterns introduced

hmmueller committed:
PathWriter - added comment to advanced article

hmmueller committed:
Merge branch 'master' of https://github.com/hmmueller/NDepCheck

hmmueller committed:
PathWriter - intermediate step to add to add arbitrary path anchors

hmmueller committed:
PathWriter - first useful implementation; Item - added DebuggerDisplayAttribute; ItemPattern is resilient against too many matchers

hmmueller committed:
Program - V.3.78

hmmueller committed:
ItemPattern - limited recognition of anywhere patterns

hmmueller committed:
PathWriter - reparied after accidental undo

hmmueller committed:
PathWriter - second implementation attempt

hmmueller committed:
FindCycleDeps, DepthFirstPathTraverser, PathWriter - reimplementation

hmmueller committed:
PathWriter - TreeWriter essentially complete (-c. options not yet tested)

hmmueller committed:
PathWriter - first correct implementaion of count options

hmmueller committed:
InputContext killed

hmmueller committed:
MarkerSet - rewritten as Dictionary<string,int> with corresponding patterns and merging behaviors.

hmmueller committed:
PathWriter - corrections for pattern recognition traversal; MarkerSet - rigidly restricting marker format; DipWriter/reader - correctly reading new MarkerSets; Sources - correctly read and written

hmmueller committed:
Option - Trim() bug fixed (relevant only for tests); CheckDeps - problem with ===> fixed and tested

hmmueller committed:
PathWriter - initial anchor can also be counting anchor

hmmueller committed:
DependencyRuleGroup - corr some comments

hmmueller committed:
Program, GlobalContext, SimpleConstraintSolver, GraphicsRenderer, PathWriter, FindCycleDeps - support for aborting; CountPattern - added operators ^and |; AbstractDepthFirstTraverser, PathWriter, PathMatch - corrected algorithm; everywhere - \r\n -> Environment.NewLine;

hmmueller committed:
IMarkerSet - public GetValue() methods; IWIthMutableMarkerSet - SetMarker(..., value) for path marking; PathWriter - temporarily renamed to OldPathWriter; ITransformers and derived classes - GetTestDependencies -> CreateSomeTestDependencies; PathFinding.* - concept of "path marking" introduced;

hmmueller committed:
FlatPathWriter - first implementation; path representation defined - dependencies along a path have markers; Dependency - ToString() shows markers; DictionaryExtensions moved to main namespace;

hmmueller committed:
FlatPathWriter - algorithm corrected (follows specification in PathSupport.cs); PathMarker - first tested implementation;

hmmueller committed:
DotNetAssemblyReader, AbstractDependencyReader - remember relative filename as "containerUri"; MarkerSet, Item, Dependency - AsFullString with length limit; CountPattern - no format checking for operands (as they can be patterns for Matcher construction); PathMarker - added simple "add-marker" option (useful for -ci); Program - number optional (default 10) for -li and -ld;

hmmueller committed:
V.3.84

hmmueller committed:
everywhere - filenames and NamedTextWriter replaced with WriteTarget (support for file appending); -li and -ld - more consistent, can write to target; everywhere - [NotNull] for GlobalContext, [NotNull, ItemNotNull] for dependencies and transformedDependencies;

hmmueller committed:
DotNetReaders - improved containerUri handling (no paths), improved location heuristics (classname + membername if sequencePoint is missing); everywhere - TextWriter replaced with TargetWriter; TargetWriter - limiting default output on Console; DOTNETCALL - renamed to DOTNETITEM; PathMarker - better formatting of indexed markers; -ld - 3-line output of dependencies; Program - exception handling for FileNotFoundException corrected; Option - filenames are not expanded to full name too early;

hmmueller committed:
-nd and -ni can check for a maximum; DependencyMatchOptions - help text corr.

hmmueller committed:
MarkCycleDeps - bug corrected which missed many cycles; somewhat better formatting of cycle markers; ItemType - somewhat better error message if no fields are defined

hmmueller committed:
Transformer - corr error message on type reading; Program - V.3.87

hmmueller committed:
Item - marker setting on creation corrected (repairs ridiculous markers in IXOS .dip files); ProjectItems - back projection works now as expected; FlatPathWriter - improved output for paths with multiple equally marked dependencies [DOES NOT YET WORK ON  LOOPBACK DEPENDENCIES]

hmmueller committed:
ProjectItems - some test for a back projection, but I cannot reproduce it

hmmueller committed:
ProjectItems - Asserts in test

hmmueller committed:
Environment - first raw implementation (without any tests :-( )

hmmueller committed:
CheckDeps, RuleViolationsWriter - better output; GlobalContext - verbose output of read files;

hmmueller committed:
Program, GlobalContext - corrected first glaring bugs in environment handling; replaced GlobalContext.Name with Environment.Name

hmmueller committed:
FlatPathWriter - better output for "broken paths"; ProjectItems - back projection now uses environments; FromTo - repaired Equals to repair bug with back projection; Program & GlobalContext & Environment - -en option now renames current environment, creation of new empty environment is now -ee;

hmmueller committed:
Preparations for introducing item+dependency factories - all new Item/new Dependency calls routed through factory at environment

hmmueller committed:
ItemType - structural mapping instead of name mapping; anonymous types and anonymous fields in types

hmmueller committed:
AnyWhereMatcher - restricted to patterns from options (but not in Projections and DependencyRules)

hmmueller committed:
DependencyMatch, ItemMatch - now have representations, which are the input patterns from which one is built; DOTNETREF and DOTNETITEM - type fields PascalCased; Environment->WorkingGraph everywhere; Dependency -  badCt >0 and questionableCt > 0 always require a not-empty reason (changed all test cases and test dependency methods); Program - GraphHide/Unhide/FiltersOption roughly implemented;

hmmueller committed:
AbstractMarkerSet - markers can contain ~, because the autogenerated markers use this for non-identifier chars; ItemType, ItemPattern - keys and subkeys of types must be non-null ("" for none); MarkDeps - added ResetBadCountOption and ResetQuestionableCountOption for convenience (ModifyDeps is too hard to write for such simple cases); PathMarker - moved to correct namespace (Transforming, not Rendering); AddPathDeps - first raw implementation; DependencyRuleSet - groups are identified by their pattern, not by their (non-unqiue) marker; FromTo - generic version added; Dependency - bug in Clone corrected, tiny change of output with new lines; GlobalContext - dependencies cleanly cloned [but copy-on-write is missing!]; DeleteGraphs deletes all matching graphs; IPathMatch - some renames; Item - Outgoing and Incoming are methods; Program, WorkingGraph - GraphHide/Unhide->GraphViewOnly/UnView to reverse logic;

hmmueller committed:
*Violations*.txt removed from repository

hmmueller committed:
* - NewItem->CreateItem; Item - .Append creates item via factory; NDpeCheck.nd - allowed Item.ctor calls defined;

hmmueller committed:
All transfomers - get WorkingGraphs for graph-spanning transformations (currently only back-projection in ProjectItems); Dependency - source presentation in violation output improved; GlobalContext - WorkingGraph handling was not correctly designed -> rewritten; Program - comments allowed in interactive input (nice for replaying -ia logs);

hmmueller committed:
WorkingGraph, Program, GlobalContext - added sticky ids, then naming better

hmmueller committed:
PathRegex - first implementation; Test* - ExcludeCoverage added

hmmueller committed:
R# green for a list of files;

hmmueller committed:
* - added ExcludeFromCodeCoverage to most ToString() methods, as they are intended only for debugging; GlobalContext - FindGraphs repaired (goes primarily on user-defined name); Program - added overview test for graph options

hmmueller committed:
PathRegex - first acceptable implementation of Advance()

hmmueller committed:
PathRegex - only reformatted; TestPathRegex - interesting test of a regex that checks whether path is multiple of 2 or 3.

hmmueller committed:
PathRegex - added some [NotNull], [CanBeNull], fixed potential NRE bug; AbstractRegexDepthFirstPathTraverser - first implementation; PathMarker - inner class renamed; RegexPathMarker - currently COMMENTED, NOT IMPLEMENTED; Ignore - public for use in tests

hmmueller committed:
RegexPathMarker - first implementation

hmmueller committed:
Regex machinery - fixed bugs, made more usable, added documentation (in HELP strings), renames, tested with (adapted) old PathMarker tests; Program - increased version to 3.90, formally only because the old PathMarker is now called OLDPathMarker, but also because so much new features are now here.

hmmueller committed:
PathRegex - help improved; FullDotNetAssemblyDependencyReader - introduced specific item types for dot net items; PathMarker - implements ITransformer; Tests - create transitive implements closures

hmmueller committed:
Transformers - rewritten with Options API (Renderers still missing)

hmmueller committed:
Renderers - replaced option strings with local option classes (API design)

hmmueller committed:
TransformerWithOptions - provided default options if no Configure() is called;  DotNetAssemblyReader - somewhat better marking of parameters from C# programs; CheckDeps - corrected naming in Options class, improved error output if no .dep search directories are provided

hmmuellerpt committed:
Optins: Better file name handling (always using full paths so that excludes work more consistently)

thoemmi commented:
@hmmueller Ich vermute, Du hast das bereits in Deinem eigenen Fork gefixt :wink:
